crates/uv-trampoline/src/bounce.rs

```diff
@@ -116,33 +125,148 @@ fn executable_filename() -> CString {
         );
 
-        exit_with_status(1);
+        let mut read_bytes = 0u32;
+
+        expect_result(
+            unsafe {
+                ReadFile(
+                    file_handle,
+                    buffer.as_mut_ptr() as *mut _,
+                    bytes_to_read,
+                    &mut read_bytes,
+                    null_mut(),
+                )
+            },
+            0,
+            || String::from("Failed to read the executable file"),
+        );
+
+        // Truncate the buffer to the actual number of bytes read.
+        buffer.truncate(read_bytes as usize);
+
+        if !buffer.ends_with(&MAGIC_NUMBER) {
+            eprintln!("Magic number 'UVUV' not found at the end of the file. Did you append the magic number, the length and the path to the python executable at the end of the file?");
+            exit_with_status(1);
+        }
+
+        // Remove the magic number
+        buffer.truncate(buffer.len() - MAGIC_NUMBER.len());
+
+        let path_len = match buffer.get(buffer.len() - PATH_LEN_SIZE..) {
+            Some(path_len) => {
+                // We only support 64bit systems where usize is guaranteed to be larger than u32.
+                u32::from_le_bytes(path_len.try_into().unwrap_or_else(|_| {
+                    eprintln!("Slice length is not equal to 4 bytes");
+                    exit_with_status(1)
+                })) as usize
+            }
+            None => {
+                eprintln!("Python executable length missing. Did you write the length of the path to the Python executable before the Magic number?");
+                exit_with_status(1);
+            }
+        };
+
+        // Remove the path length
+        buffer.truncate(buffer.len() - PATH_LEN_SIZE);
+
+        if let Some(path_offset) = buffer.len().checked_sub(path_len) {
+            buffer.drain(..path_offset);
+            buffer.push(b'\0');
+
+            break CString::from_vec_with_nul(buffer).unwrap_or_else(|_| {
+                eprintln!("Python executable path is not correctly null terminated.");
+                exit_with_status(1)
+            });
+        } else {
+            let new_len = path_len
+                .saturating_add(MAGIC_NUMBER.len())
+                .saturating_add(PATH_LEN_SIZE);
+
+            if path_len >= usize::try_from(file_size).unwrap_or(usize::MAX) {
+                eprintln!("The length of the path to the Python executable is larger than the file size. Did you write the length of the path to the Python executable before the Magic number?");
+                exit_with_status(1);
+            }
+
+            buffer.resize(new_len, 0);
+        }
     };
 
-    let mut buffer = bytes[..last_backslash + 1].to_vec();
-    buffer.extend_from_slice(if is_gui {
-        b"pythonw.exe"
-    } else {
-        b"python.exe"
+    expect_result(unsafe { CloseHandle(file_handle) }, 0, || {
+        String::from("Failed to close file handle")
```

Comment:
OK, so reading the above, I think it works like this:

* It first tries to read 1KB from the end of a file.
* If it finds the magic number, path length and path all within that 1KB, it stops.
* Otherwise, it resizes the buffer's capacity to whatever the decoded `path_len` is.
* It then goes back to the first step again, but read `{path_len}KB` instead.

I think there might be a couple issues with this approach, both of which center around trusting `path_len`:

1. If `path_len` is `u32::MAX`, then this will allocate 4GB.
2. If the file is being mutated while we're doing this in a very specific way, then it's possible this loop isn't guaranteed to terminate.

(2) is kind of a stretch, but (1) I think is a real enough issue.

My suggestion here would be to read 4KB from the end of the file, and if you can't find the path in there, consider it malformed and give up.

Reply:
Thanks for the thorough review. I'll carefully go over the implementation again tomorrow. 

2. shouldn't be a problem because [we open the file](https://github.com/astral-sh/puffin/blob/e772c05b88019319f57cd52b084666291c2b262e/crates/uv-trampoline/src/bounce.rs#L147-L155) with [`FILE_SHARE_READ`](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilea#parameters) only, prohibiting other processes from mutating or deleting the file while we're using it. 

Not trusting `path_len` makes sense as well does limiting. I'll probably go with a higher default, e.g. even URl have an [upper limit of 2MB](https://chromium.googlesource.com/chromium/src/+/master/docs/security/url_display_guidelines/url_display_guidelines.md#URL-Length). Agreed, URLs may contain application data which is different but I don't think it hurts to go somewhat higher than 4KB.

Reply:
From https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry:

> The Windows API has many functions that also have Unicode versions to permit an extended-length path for a maximum total path length of 32,767 characters.
> The maximum path of 32,767 characters is approximate, because the "\\?\" prefix may be expanded to a longer string by the system at run time, and this expansion applies to the total length.

Reply:
So my read of that doc is that there isn't a pre-defined upper bound on how long a path can be. Does that match yours? If that's true, I still think we need to place some kind of reasonable upper bound on what we're willing to accept. I think 4KB is probably sufficient, but 32KB would be fine with me too.

Reply:
I limited it to 32KB